IllegalAccessError trying to run UberJar when public class C extends public class B extends package class A

My class inheritance looks like this:

class Apackage.A
public class Apackage.B extends A
public class Bpackage.C extends B

This has been working up to now with Wildfly/Thorntail, etc.
When trying to run an UberJar quarkus built, I'm getting:

java.lang.IllegalAccessError: tried to access class A from class C_Bean

I couldn't find any good documentation on transitive class access though, so maybe this is not a bug as such. Also have no idea how C_Bean comes into life (presumably already including all of B's funtionality, therefore stripping access to B and directly accessing A?)

I cannot easily change Apackage.A to have public access as that comes from a library.

@Xnyle, thank you for those details. I was able to reproduce this based on what you provided.

@stuartwdouglas, this now looks like an Arc issue rather than classloading or uber jar related. I've a simple testcase which reproduces this against upstream master. The test case is here https://github.com/jaikiran/quarkus/commit/6da288cfbdaa2b65acae9619446e4b47fb6a028f.

This test fails:
```
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:310)
	... 42 more
Caused by: java.lang.IllegalAccessError: tried to access class io.quarkus.resteasy.test.cdi.internal.Hello from class io.quarkus.resteasy.test.cdi.MorningGreeting_Bean
	at io.quarkus.resteasy.test.cdi.MorningGreeting_Bean.<init>(MorningGreeting_Bean.zig:84)
	at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Default_ComponentsProvider.zig:153)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:38)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:102)
	at io.quarkus.arc.Arc.initialize(Arc.java:19)
	at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:34)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy_0(ArcProcessor$generateResources15.zig:72)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy(ArcProcessor$generateResources15.zig:36)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:280)
	... 42 more

```
I ended up using `dumpClass` to dump the generated bean class for `MorningGreeting_Bean` (the one which is trying to access the package private `Hello`) and here's its content:

```
Classfile /private/tmp/code-with-quarkus/target/test-classes/io/quarkus/resteasy/test/cdi/MorningGreeting_Bean.class
  Last modified 11 Nov, 2019; size 1985 bytes
  MD5 checksum c97c0b25942aaa6a7aa96c16ce8873e2
  Compiled from "MorningGreeting_Bean.zig"
public class io.quarkus.resteasy.test.cdi.MorningGreeting_Bean implements io.quarkus.arc.InjectableBean,java.util.function.Supplier
  minor version: 0
  major version: 52
  flags: ACC_PUBLIC, ACC_SUPER, ACC_SYNTHETIC
Constant pool:
   #1 = Utf8               io/quarkus/resteasy/test/cdi/MorningGreeting_Bean
   #2 = Class              #1             // io/quarkus/resteasy/test/cdi/MorningGreeting_Bean
   #3 = Utf8               java/lang/Object
   #4 = Class              #3             // java/lang/Object
   #5 = Utf8               io/quarkus/arc/InjectableBean
   #6 = Class              #5             // io/quarkus/arc/InjectableBean
   #7 = Utf8               java/util/function/Supplier
   #8 = Class              #7             // java/util/function/Supplier
   #9 = Utf8               MorningGreeting_Bean.zig
  #10 = Utf8               types
  #11 = Utf8               Ljava/util/Set;
  #12 = Utf8               create
  #13 = Utf8               (Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
  #14 = Utf8               io/quarkus/resteasy/test/cdi/MorningGreeting
  #15 = Class              #14            // io/quarkus/resteasy/test/cdi/MorningGreeting
  #16 = Utf8               <init>
  #17 = Utf8               ()V
  #18 = NameAndType        #16:#17        // "<init>":()V
  #19 = Methodref          #15.#18        // io/quarkus/resteasy/test/cdi/MorningGreeting."<init>":()V
  #20 = Utf8               getBeanClass
  #21 = Utf8               ()Ljava/lang/Class;
  #22 = Utf8               (Ljavax/enterprise/context/spi/CreationalContext;)Ljava/lang/Object;
  #23 = NameAndType        #12:#13        // create:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
  #24 = Methodref          #2.#23         // io/quarkus/resteasy/test/cdi/MorningGreeting_Bean.create:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
  #25 = Methodref          #4.#18         // java/lang/Object."<init>":()V
  #26 = Utf8               java/util/HashSet
  #27 = Class              #26            // java/util/HashSet
  #28 = Methodref          #27.#18        // java/util/HashSet."<init>":()V
  #29 = Utf8               java/util/Set
  #30 = Class              #29            // java/util/Set
  #31 = Utf8               io/quarkus/resteasy/test/cdi/internal/PublicHello
  #32 = Class              #31            // io/quarkus/resteasy/test/cdi/internal/PublicHello
  #33 = Utf8               add
  #34 = Utf8               (Ljava/lang/Object;)Z
  #35 = NameAndType        #33:#34        // add:(Ljava/lang/Object;)Z
  #36 = InterfaceMethodref #30.#35        // java/util/Set.add:(Ljava/lang/Object;)Z
  #37 = Utf8               io/quarkus/resteasy/test/cdi/internal/Hello
  #38 = Class              #37            // io/quarkus/resteasy/test/cdi/internal/Hello
  #39 = Utf8               io/quarkus/resteasy/test/cdi/Greeting
  #40 = Class              #39            // io/quarkus/resteasy/test/cdi/Greeting
  #41 = Utf8               java/util/Collections
  #42 = Class              #41            // java/util/Collections
  #43 = Utf8               unmodifiableSet
  #44 = Utf8               (Ljava/util/Set;)Ljava/util/Set;
  #45 = NameAndType        #43:#44        // unmodifiableSet:(Ljava/util/Set;)Ljava/util/Set;
  #46 = Methodref          #42.#45        // java/util/Collections.unmodifiableSet:(Ljava/util/Set;)Ljava/util/Set;
  #47 = NameAndType        #10:#11        // types:Ljava/util/Set;
  #48 = Fieldref           #2.#47         // io/quarkus/resteasy/test/cdi/MorningGreeting_Bean.types:Ljava/util/Set;
  #49 = Utf8               get
  #50 = Utf8               ()Ljava/lang/Object;
  #51 = Utf8               getIdentifier
  #52 = Utf8               ()Ljava/lang/String;
  #53 = Utf8               1d02a4fbe21970dc98b22704447f3b7369f2f41e
  #54 = String             #53            // 1d02a4fbe21970dc98b22704447f3b7369f2f41e
  #55 = NameAndType        #49:#13        // get:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
  #56 = Methodref          #2.#55         // io/quarkus/resteasy/test/cdi/MorningGreeting_Bean.get:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
  #57 = Utf8               io/quarkus/arc/impl/CreationalContextImpl
  #58 = Class              #57            // io/quarkus/arc/impl/CreationalContextImpl
  #59 = Utf8               addDependencyToParent
  #60 = Utf8               (Lio/quarkus/arc/InjectableBean;Ljava/lang/Object;Ljavax/enterprise/context/spi/CreationalContext;)V
  #61 = NameAndType        #59:#60        // addDependencyToParent:(Lio/quarkus/arc/InjectableBean;Ljava/lang/Object;Ljavax/enterprise/context/spi/CreationalContext;)V
  #62 = Methodref          #58.#61        // io/quarkus/arc/impl/CreationalContextImpl.addDependencyToParent:(Lio/quarkus/arc/InjectableBean;Ljava/lang/Object;Ljavax/enterprise/context/spi/CreationalContext;)V
  #63 = Utf8               getTypes
  #64 = Utf8               ()Ljava/util/Set;
  #65 = Utf8               Code
  #66 = Utf8               LineNumberTable
  #67 = Utf8               SourceFile
{
  public io.quarkus.resteasy.test.cdi.MorningGreeting create(javax.enterprise.context.spi.CreationalContext);
    descriptor: (Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=2, args_size=2
         0: new           #15                 // class io/quarkus/resteasy/test/cdi/MorningGreeting
         3: dup
         4: invokespecial #19                 // Method io/quarkus/resteasy/test/cdi/MorningGreeting."<init>":()V
         7: areturn
      LineNumberTable:
        line 20: 0
        line 21: 3
        line 23: 4
        line 24: 7

  public java.lang.Class getBeanClass();
    descriptor: ()Ljava/lang/Class;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: ldc           #15                 // class io/quarkus/resteasy/test/cdi/MorningGreeting
         2: areturn
      LineNumberTable:
        line 35: 0
        line 36: 2

  public java.lang.Object create(javax.enterprise.context.spi.CreationalContext);
    descriptor: (Ljavax/enterprise/context/spi/CreationalContext;)Ljava/lang/Object;
    flags: ACC_PUBLIC, ACC_BRIDGE
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokevirtual #24                 // Method create:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
         5: areturn
      LineNumberTable:
        line 47: 0
        line 48: 1
        line 50: 2
        line 51: 5

  public io.quarkus.resteasy.test.cdi.MorningGreeting_Bean();
    descriptor: ()V
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=3, args_size=1
         0: aload_0
         1: invokespecial #25                 // Method java/lang/Object."<init>":()V
         4: new           #27                 // class java/util/HashSet
         7: dup
         8: invokespecial #28                 // Method java/util/HashSet."<init>":()V
        11: astore_1
        12: aload_1
        13: checkcast     #30                 // class java/util/Set
        16: ldc           #32                 // class io/quarkus/resteasy/test/cdi/internal/PublicHello
        18: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        23: pop
        24: aload_1
        25: checkcast     #30                 // class java/util/Set
        28: ldc           #15                 // class io/quarkus/resteasy/test/cdi/MorningGreeting
        30: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        35: pop
        36: aload_1
        37: checkcast     #30                 // class java/util/Set
        40: ldc           #38                 // class io/quarkus/resteasy/test/cdi/internal/Hello
        42: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        47: pop
        48: aload_1
        49: checkcast     #30                 // class java/util/Set
        52: ldc           #40                 // class io/quarkus/resteasy/test/cdi/Greeting
        54: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        59: pop
        60: aload_1
        61: checkcast     #30                 // class java/util/Set
        64: ldc           #4                  // class java/lang/Object
        66: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        71: pop
        72: aload_1
        73: checkcast     #30                 // class java/util/Set
        76: invokestatic  #46                 // Method java/util/Collections.unmodifiableSet:(Ljava/util/Set;)Ljava/util/Set;
        79: astore_2
        80: aload_0
        81: aload_2
        82: putfield      #48                 // Field types:Ljava/util/Set;
        85: return
      LineNumberTable:
        line 62: 0
        line 64: 1
        line 65: 4
        line 66: 7
        line 68: 8
        line 69: 11
        line 70: 12
        line 71: 13
        line 72: 16
        line 74: 18
        line 75: 23
        line 76: 24
        line 77: 25
        line 78: 28
        line 80: 30
        line 81: 35
        line 82: 36
        line 83: 37
        line 84: 40
        line 86: 42
        line 87: 47
        line 88: 48
        line 89: 49
        line 90: 52
        line 92: 54
        line 93: 59
        line 94: 60
        line 95: 61
        line 96: 64
        line 98: 66
        line 99: 71
        line 100: 72
        line 101: 73
        line 103: 76
        line 104: 79
        line 105: 80
        line 106: 81
        line 108: 82
        line 109: 85

  public java.lang.Object get();
    descriptor: ()Ljava/lang/Object;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: areturn
      LineNumberTable:
        line 120: 0
        line 121: 1

  public java.lang.String getIdentifier();
    descriptor: ()Ljava/lang/String;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: ldc           #54                 // String 1d02a4fbe21970dc98b22704447f3b7369f2f41e
         2: areturn
      LineNumberTable:
        line 132: 0
        line 133: 2

  public java.lang.Object get(javax.enterprise.context.spi.CreationalContext);
    descriptor: (Ljavax/enterprise/context/spi/CreationalContext;)Ljava/lang/Object;
    flags: ACC_PUBLIC, ACC_BRIDGE
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokevirtual #56                 // Method get:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
         5: areturn
      LineNumberTable:
        line 144: 0
        line 145: 1
        line 147: 2
        line 148: 5

  public io.quarkus.resteasy.test.cdi.MorningGreeting get(javax.enterprise.context.spi.CreationalContext);
    descriptor: (Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
    flags: ACC_PUBLIC
    Code:
      stack=3, locals=3, args_size=2
         0: aload_0
         1: aload_1
         2: invokevirtual #24                 // Method create:(Ljavax/enterprise/context/spi/CreationalContext;)Lio/quarkus/resteasy/test/cdi/MorningGreeting;
         5: astore_2
         6: aload_0
         7: checkcast     #6                  // class io/quarkus/arc/InjectableBean
        10: aload_2
        11: aload_1
        12: invokestatic  #62                 // Method io/quarkus/arc/impl/CreationalContextImpl.addDependencyToParent:(Lio/quarkus/arc/InjectableBean;Ljava/lang/Object;Ljavax/enterprise/context/spi/CreationalContext;)V
        15: aload_2
        16: areturn
      LineNumberTable:
        line 159: 0
        line 160: 1
        line 162: 2
        line 163: 5
        line 164: 6
        line 165: 7
        line 166: 10
        line 167: 11
        line 169: 12
        line 170: 15
        line 171: 16

  public java.util.Set getTypes();
    descriptor: ()Ljava/util/Set;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: getfield      #48                 // Field types:Ljava/util/Set;
         4: areturn
      LineNumberTable:
        line 182: 0
        line 184: 1
        line 185: 4
}
SourceFile: "MorningGreeting_Bean.zig"
```
Notice:
```
public io.quarkus.resteasy.test.cdi.MorningGreeting_Bean();
    descriptor: ()V
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=3, args_size=1
         0: aload_0
         1: invokespecial #25                 // Method java/lang/Object."<init>":()V
         4: new           #27                 // class java/util/HashSet
         7: dup
         8: invokespecial #28                 // Method java/util/HashSet."<init>":()V
        11: astore_1
        12: aload_1
        13: checkcast     #30                 // class java/util/Set
        16: ldc           #32                 // class io/quarkus/resteasy/test/cdi/internal/PublicHello
        18: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        23: pop
        24: aload_1
        25: checkcast     #30                 // class java/util/Set
        28: ldc           #15                 // class io/quarkus/resteasy/test/cdi/MorningGreeting
        30: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
        35: pop
        36: aload_1
        37: checkcast     #30                 // class java/util/Set
        40: ldc           #38                 // class io/quarkus/resteasy/test/cdi/internal/Hello
        42: invokeinterface #36,  2           // InterfaceMethod java/util/Set.add:(Ljava/lang/Object;)Z
```
So the constructor of the generated bean is adding this (package private) class `Hello` into it `Set` to keep track of the `types` exposed by this bean.

So this boils down to Arc iterating over a bean and then trying to find the "types" it exposes. I think it's all in https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java#L256 where it iterates the super classes. We might need @mkouba's help/inputs here.

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/BeanGenerationTest.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/BeanGenerationTest.java
@@ -1,0 +1,32 @@
+package io.quarkus.resteasy.test.cdi;
+
+import static io.restassured.RestAssured.when;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.resteasy.test.cdi.internal.PublicHello;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ *
+ */
+public class BeanGenerationTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addPackage(PublicHello.class.getPackage())
+                    .addClasses(Greeting.class, MorningGreeting.class, GreetingEndpoint.class));
+
+    @Test
+    public void testInvocation() throws Exception {
+        when().get("/cdi-greeting/greet")
+                .then()
+                .statusCode(200)
+                .body(Matchers.is("Good Morning"));
+    }
+}

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/Greeting.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/Greeting.java
@@ -1,0 +1,14 @@
+package io.quarkus.resteasy.test.cdi;
+
+import io.quarkus.resteasy.test.cdi.internal.PublicHello;
+
+/**
+ *
+ */
+public class Greeting extends PublicHello {
+
+    @Override
+    public String greet() {
+        return "Hello user";
+    }
+}

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/GreetingEndpoint.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/GreetingEndpoint.java
@@ -1,0 +1,21 @@
+package io.quarkus.resteasy.test.cdi;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+/**
+ *
+ */
+@Path("/cdi-greeting")
+public class GreetingEndpoint {
+
+    @Inject
+    Greeting greeting;
+
+    @GET
+    @Path("/greet")
+    public String greet() {
+        return greeting.greet();
+    }
+}

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/MorningGreeting.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/MorningGreeting.java
@@ -10,6 +10,6 @@ public class MorningGreeting extends Greeting {
 
     @Override
     public String greet() {
-        return "Good morning";
+        return "Good Morning";
     }
 }

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/MorningGreeting.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/MorningGreeting.java
@@ -1,0 +1,15 @@
+package io.quarkus.resteasy.test.cdi;
+
+import javax.enterprise.context.Dependent;
+
+/**
+ *
+ */
+@Dependent
+public class MorningGreeting extends Greeting {
+
+    @Override
+    public String greet() {
+        return "Good morning";
+    }
+}

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/internal/Hello.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/internal/Hello.java
@@ -1,0 +1,11 @@
+package io.quarkus.resteasy.test.cdi.internal;
+
+/**
+ *
+ */
+class Hello {
+
+    void noop() {
+        System.out.println("Hello");
+    }
+}

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/internal/PublicHello.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/cdi/internal/PublicHello.java
@@ -1,0 +1,11 @@
+package io.quarkus.resteasy.test.cdi.internal;
+
+/**
+ *
+ */
+public class PublicHello extends Hello {
+
+    public String greet() {
+        return "Hello";
+    }
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -1,6 +1,7 @@
 package io.quarkus.arc.processor;
 
 import static io.quarkus.arc.processor.IndexClassLookupUtils.getClassByName;
+import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
 import io.quarkus.arc.impl.GenericArrayTypeImpl;
 import io.quarkus.arc.impl.ParameterizedTypeImpl;
@@ -49,7 +50,8 @@ final class Types {
 
     static ResultHandle getTypeHandle(BytecodeCreator creator, Type type) {
         if (Kind.CLASS.equals(type.kind())) {
-            return creator.loadClass(type.asClassType().name().toString());
+            String className = type.asClassType().name().toString();
+            return doLoadClass(creator, className);
         } else if (Kind.TYPE_VARIABLE.equals(type.kind())) {
             // E.g. T -> new TypeVariableImpl("T")
             TypeVariable typeVariable = type.asTypeVariable();
@@ -79,7 +81,7 @@ final class Types {
             return creator.newInstance(
                     MethodDescriptor.ofConstructor(ParameterizedTypeImpl.class, java.lang.reflect.Type.class,
                             java.lang.reflect.Type[].class),
-                    creator.loadClass(parameterizedType.name().toString()), typeArgsHandle);
+                    doLoadClass(creator, parameterizedType.name().toString()), typeArgsHandle);
 
         } else if (Kind.ARRAY.equals(type.kind())) {
             Type componentType = type.asArrayType().component();
@@ -126,6 +128,18 @@ final class Types {
         } else {
             throw new IllegalArgumentException("Unsupported bean type: " + type.kind() + ", " + type);
         }
+    }
+
+    private static ResultHandle doLoadClass(BytecodeCreator creator, String className) {
+        //we need to use Class.forName as the class may be package private
+        ResultHandle currentThread = creator
+                .invokeStaticMethod(ofMethod(Thread.class, "currentThread", Thread.class));
+        ResultHandle tccl = creator.invokeVirtualMethod(
+                ofMethod(Thread.class, "getContextClassLoader", ClassLoader.class),
+                currentThread);
+        return creator.invokeStaticMethod(
+                ofMethod(Class.class, "forName", Class.class, String.class, boolean.class, ClassLoader.class),
+                creator.load(className), creator.load(true), tccl);
     }
 
     static Type getProviderType(ClassInfo classInfo) {

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java
@@ -16,10 +16,15 @@ public final class Arc {
 
     public static ArcContainer initialize() {
         if (INITIALIZED.compareAndSet(false, true)) {
-            ArcContainerImpl container = new ArcContainerImpl();
-            INSTANCE.set(container);
-            container.init();
-            return container;
+            try {
+                ArcContainerImpl container = new ArcContainerImpl();
+                INSTANCE.set(container);
+                container.init();
+                return container;
+            } catch (Throwable t) {
+                INITIALIZED.set(false);
+                throw new RuntimeException("Failed to initialize Arc", t);
+            }
         }
         return container();
     }

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
@@ -301,7 +301,7 @@ public class QuarkusUnitTest
 
                 Object actualTest = factory.get();
                 extensionContext.getStore(ExtensionContext.Namespace.GLOBAL).put(testClass.getName(), actualTest);
-            } catch (Exception e) {
+            } catch (Throwable e) {
                 started = false;
                 if (assertException != null) {
                     if (e instanceof RuntimeException) {

